include/libpmemobj++/experimental/array.hpp

```diff
@@ -0,0 +1,671 @@
+						       N, snapshot_size)};
+	}
+
+	/**
+	 * Returns const slice.
+	 *
+	 * @param[in] start start index of requested range
+	 * @param[in] n number of elements in range
+	 * @return slice from start to start + n
+	 */
+	slice<const_iterator>
+	range(size_type start, size_type n) const
+	{
+		if (start + n >= N)
+			throw std::out_of_range("array::slice");
+
+		return {const_iterator(_data + start),
+			const_iterator(_data + start + n)};
+	}
+
+	/**
+	 * Returns const slice.
+	 *
+	 * @param[in] start start index of requested range
+	 * @param[in] n number of elements in range
+	 * @return slice from start to start + n
+	 */
+	slice<const_iterator>
+	crange(size_type start, size_type n) const
+	{
+		if (start + n >= N)
+			throw std::out_of_range("array::slice");
+
+		return {const_iterator(_data + start),
+			const_iterator(_data + start + n)};
+	}
+
+	/**
+	 * Returns size of the array.
+	 */
+	constexpr size_type
+	size() const noexcept
+	{
+		return N;
+	}
+
+	/**
+	 * Returns the maximum size of the array.
+	 */
+	constexpr size_type
+	max_size() const noexcept
+	{
+		return N;
+	}
+
+	/**
+	 * Checks wheter array is empty.
+	 */
+	constexpr bool
+	empty() const noexcept
+	{
+		return size() != 0;
+	}
+
+	/**
+	 * Adds entire array to a transaction and fills array with
+	 * specified value.
+	 *
+	 * @throw transaction_error when adding the object to the
+	 *		transaction failed.
+	 */
+	void
+	fill(const_reference value)
+	{
+		detail::conditional_add_to_tx(this);
+		std::fill(_data, _data + size(), value);
+	}
+
+	/**
+	 * Swaps content with other array's content.
+	 * Adds both arrays to a transaction.
+	 *
+	 * @throw transaction_error when adding the object to the
+	 *		transaction failed.
+	 */
+	void
+	swap(array &other)
+	{
+		detail::conditional_add_to_tx(this);
+		detail::conditional_add_to_tx(&other);
+
+		std::swap_ranges(_data, _data + size(), other._data);
+	}
+};
+
+template <typename T, std::size_t N>
+inline bool
+operator==(const array<T, N> &lhs, const array<T, N> &rhs)
+{
+	return std::equal(lhs.cbegin(), lhs.cend(), rhs.cbegin());
+}
+
+template <typename T, std::size_t N>
+inline bool
+operator!=(const array<T, N> &lhs, const array<T, N> &rhs)
+{
+	return !(lhs == rhs);
+}
+
+template <typename T, std::size_t N>
+inline bool
+operator<(const array<T, N> &lhs, const array<T, N> &rhs)
+{
+	return std::lexicographical_compare(lhs.cbegin(), lhs.cend(),
+					    rhs.cbegin(), rhs.cend());
+}
+
+template <typename T, std::size_t N>
+inline bool
+operator>(const array<T, N> &lhs, const array<T, N> &rhs)
+{
+	return rhs < lhs;
+}
+
+template <typename T, std::size_t N>
+inline bool
+operator>=(const array<T, N> &lhs, const array<T, N> &rhs)
+{
+	return !(lhs < rhs);
+}
+
+template <typename T, std::size_t N>
+inline bool
+operator<=(const array<T, N> &lhs, const array<T, N> &rhs)
+{
+	return !(lhs > rhs);
+}
+
+} /* namespace experimental */
+
+} /* namespace obj */
+
+} /* namespace pmem */
+
+namespace std
```

Comment:
All specializations should be in the same namespace as the class IMHO. That's the approach we've taken in `persistent_ptr` and other classes.